### PR TITLE
Fix JsonDocument thread safety.

### LINKED ISSUE #76440: JsonDocument.GetString implementation is not thread safe
Even though `JsonDocument` is nominally immutable (modulo disposability), the internal `GetString` implementation employs a caching scheme that is not thread safe:

https://github.com/dotnet/runtime/blob/5c8aade5384cbad2d086e7fae482ba0b692d3601/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs#L30

Since the cached field is a struct that cannot be written to or read from atomically. It might be that we could fix this in the code using a regular `System.Tuple`, thus ensuring atomic access to cached entries. Obviously this has the downside of incurring one additional allocation, but at least we can guarantee immutability (modulo disposability of course).

Related to https://github.com/dotnet/runtime/issues/17975

cc @layomia @krwq

_Originally posted by @eiriktsarpalis in https://github.com/dotnet/docs/issues/20563#issuecomment-1263202217_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Removes the cache breaking JsonDocument's thread safety.

Fixes https://github.com/dotnet/runtime/issues/76440. Supersedes #76450.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonDocumentTests.cs [hunk 1]
```diff
@@ -2744,11 +2745,9 @@ public static void ObjectEnumeratorIndependentWalk()
                     Assert.Equal(test, property.Value.GetInt32());
                     test++;
 
-                    // Subsequent read of the same JsonProperty doesn't allocate a new string
-                    // (if another property is inspected from the same document that guarantee
-                    // doesn't hold).
+                    // Subsequent read of the same JsonProperty should return an equal string
                     string propertyName2 = property.Name;
-                    Assert.Same(propertyName, propertyName2);
```
Comment (krwq) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonDocumentTests.cs:2751]: Do we know why this was optimized in the first place?

**Review by krwq — APPROVED** [review]
Code-wise looks good, please validate if we care about perf of reading same property multiple times (I personally prefer cleaner code to micro-optimizations but we should validate this won't regress perf of some important scenario we might be missing)

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonDocumentTests.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonDocumentTests.cs:2751]: See https://github.com/dotnet/runtime/pull/76450#issuecomment-1264839881

**Pull request merged**